Cleanup

rhaschke opened the pull request:
Dear @guihomework,
Trying to understand `pwlf`, I started first looking into your multi-calib commits. I have collected some required changes in this PR.
Please have a look into the individual commits and the review comments I (will) add via github.

As far as I understand the basic idea was to allow different calibrations for different taxel subsets, which will be very useful.
I have some suggestions regarding the semantics and yaml syntax to further improve:
* Why do you use the main key `calib`? Will there be any other key? My suggestion is to drop it and just use a sequence.
* The `sensor_name` is not used yet. Do we need it at all? Is that just for documentation?
* As far as I understand, `single_calib_` is used for all taxels if:
  * there is no multi-calib given at all
  * there is no `idx_range` given for any sequence item
  * there is a single, `negative` idx specified in `idx_range` for any sequence item
    I suggest to drop this last condition or replace it by `idx_range: default`. 
    My key suggestion is to use this as the default (or fallback) calib, if nothing more specific was defined for a taxel. Currently, `single_calib_` is only actually used when `calib_` is empty and calib_ is only used if it defines calibs for all taxel inputs.
    If we agree on this new semantics, I would cleanup the code some more towards this direction.

Hence, an example yaml config would look like this:
```yaml
  - sensor_name: board0
    type: PWL
    idx_range: [[0, 50], [51,90]]
    values:
      0:    0
      1024: 10
      4095: 4095
  - sensor_name: board0
    type: PWL
    idx_range: default
    values:
      0:    -3.0
      4095: -5.0
```
What do you think?

rhaschke commented on tactile_calibration/tactile_state_calibrator/CMakeLists.txt:
```diff
@@ -17,16 +14,10 @@ catkin_package(
 
 add_executable(${PROJECT_NAME} src/tactile_state_calibrator.cpp)
-target_link_libraries(${PROJECT_NAME} PUBLIC ${catkin_LIBRARIES} tactile_filters PRIVATE ${YAML_LIBRARIES})
+target_link_libraries(${PROJECT_NAME} ${catkin_LIBRARIES} tactile_filters ${YAML_LIBRARIES})
```
For an executable PUBLIC and PRIVATE doesn't make much sense. It will not be used as a dependency for some other target.

guihomework replied:
lack of knowledge of PUBLIC / PRIVATE that I copied from a library indeed. Copying you is not sufficient, understand is more important, and I tried understand it but did not understand it. Now it is clear it was a mistake.

rhaschke commented on tactile_calibration/tactile_state_calibrator/src/tactile_state_calibrator.cpp:
```diff
@@ -325,14 +323,8 @@ TactileStateCalibrator::~TactileStateCalibrator()
-  {
-    for (auto p : calibs_)
-    {
-      delete p;
```
As the same calib pointer is stored for several taxels, this delete will multiply free the memory!

guihomework replied:
thanks for spotting that double free.
what happened, is an incomplete optimization session. At first the calibration were duplicates.

guihomework commented:
This PR is annoying as it might break something that is in production.
it would be better to first discuss that

* Why do you use the main key calib? 
 because I was asked to keep the code compatible with the old style for one caib-fo-all which was a yaml::map as is. Detecting the different types was easier with a starting tag. 
* Will there be any other key?
 did not think about that. Currently not.

> My suggestion is to drop it and just use a sequence 

Dropping means changing the structure of the yaml which will break the code at our customers. This should be discussed if it is absolutely necessary now. And if yes, synchronize with them to do the switch.

Personallu I find the yaml easier to read if there is a top keyword, similar to the "<robot>" tag in all our xacros... although they are not robots.

* The sensor_name is not used yet. Do we need it at all? Is that just for documentation?

this is all because I had to plan ahead to provide code to customers but permit future extensions.
We have incoming tactile_messages which contain several sensors with different names and different vector size. One must distinguish the calibration for those in the future. I would not say it is not "used" yet, I would say it is not "handled" yet, to have a 80/20 rule and get the basic function running, but avoid needing to change the calibration file again.

> As far as I understand, single_calib_ is used for all taxels if:
    there is no multi-calib given at all

yes if the old type of yaml calib is given (back ward compatibility with the code at our customer

>    there is no idx_range given for any sequence item

yes, this was the new default first idea, but was creating issues in some of the tools that did not handle so well empty sequences, so I think I wanted to switch to a negative value.

>     there is a single, negative idx specified in idx_range for any sequence item

I think I did not clarify this case yet.  

>     I suggest to drop this last condition or replace it by idx_range: default.

The problem are that making this a string and not a vector of scala/vectors required to handle more switch cases and parsing than if it is always a vector.
I see that it is could be easier for the human (but then more complex for the developer), but I believe some humans (me) prefer to have an empty vector or some specific value telling "all" than having a change of type.

>    My key suggestion is to use this as the default (or fallback) calib, if nothing more specific was defined for a taxel. Currently, single_calib_ is only actually used when calib_ is empty and calib_ is only used if it defines calibs for all taxel inputs.

I need to process far more cases that we have in all our hardware to confirm what you say. I did it but now I doubt.


>     If we agree on this new semantics, I would cleanup the code some more towards this direction.

please no cleanup until we have a meeting.

rhaschke commented:
> The idea is to NOT require YAML, but I can see that I did not use #ifdefs everywhere. (one function needing a yaml node should be protected as well. This should be fixed, not requiring yaml

Essentially, the whole code doesn't work if yaml is not available, because file parsing requires yaml.

rhaschke commented:
> I would prefer to keep the `verbose` as it was.

Is this really needed by the customer or was that a debugging feature?
As you said, introducing `verbose` was a strange choice, not following the standards. IMHO it makes the code uglier and unnecessarily complex. So, I still suggest to remove it. As a compromise, we could easily configure DEBUG level logging using the `verbose` ROS parameter. Yes, this still will drop the two-level verbosity.

guihomework commented:
I have no idea how to respond to single answers to my inline comments, so it will be here

Then let's fix the code to work without YAML. It is made optional in tactile_filters, it must work here too. It is apparently not covered yet. My mistake, I know how this code was produced at first, in a rush due to bad time planning. so this was not done properly and I should fix it. 
This would be a new issue to report to make the code work without YAML.

guihomework commented:
> > I would prefer to keep the `verbose` as it was.
> 
> Is this really needed by the customer or was that a debugging feature?
> As you said, introducing `verbose` was a strange choice, not following the standards. IMHO it makes the code uglier and unnecessarily complex. So, I still suggest to remove it. As a compromise, we could easily configure DEBUG level logging using the `verbose` ROS parameter. Yes, this still will drop the two-level verbosity.

apparently one can quote reply which copies things... ok.

The customer will not debug, I will debug with what the customer sends as error/log. So it must be simple for the customer to change verbosity and also not flood for basic debugging.  If it is simple to change verbosity without making launch files uggly, maybe this a compromise

guihomework commented on tactile_calibration/tactile_state_calibrator/src/tactile_state_calibrator.cpp:
```diff
@@ -85,9 +86,9 @@ bool TactileStateCalibrator::extract_idx_range(const YAML::Node &node, std::map<
               ROS_DEBUG_STREAM("  index range valid" );
               // add the range to the idx_map and associate it to the calibration pointer
-              for (unsigned int i = idx_ranges[0]; i <= idx_ranges[1];++i) 
+              for (int i = idx_ranges[0]; i <= idx_ranges[1]; ++i)
```
If I am not mistaken i  can never be negative in this filling up case, so why lose half of the potential range of the int ?

rhaschke replied:
Change to unsigned int everywhere.

guihomework commented on tactile_calibration/tactile_state_calibrator/src/tactile_state_calibrator.h:
```diff
@@ -43,11 +43,12 @@ class TactileStateCalibrator
+  bool fill_calibs(const std::map<unsigned int, std::shared_ptr<tactile::Calibration> > &map);
   /**
-   * extract index of taxels a calib applies to
+   * extract indeces of taxels a calib applies to
```
if we want to fix, let's use either  indexes or indices. I prefer the latter.

rhaschke replied:
```suggestion
   * extract indices of taxels a calib applies to
```

guihomework commented on tactile_calibration/tactile_calibration_tools/CMakeLists.txt:
```diff
@@ -104,7 +104,7 @@ catkin_python_setup()
 #  INCLUDE_DIRS include
 #  LIBRARIES tactile_calibration_tools
-#  CATKIN_DEPENDS rosbag rospy tactile_msgs
+CATKIN_DEPENDS tactile_msgs
```
could we do a separate catkin_lint cleanup campain on the whole repo and not put that in this single PR ? there are many many errors in this repo with catkin_lint (your and my code) and this is maybe worth to do separately ?

guihomework commented on tactile_calibration/tactile_calibration_tools/src/tactile_calibration_tools/calibration_utils.py:
```diff
@@ -5,7 +5,6 @@
 import matplotlib.pyplot as plt
 from datetime import datetime
 import os
-#pip install pwlf
```
Could we not keep that ? it was not left there by mistake. It helps co-workers...

rhaschke commented on tactile_calibration/tactile_state_calibrator/src/tactile_state_calibrator.cpp:
```diff
@@ -114,7 +114,7 @@ bool TactileStateCalibrator::extract_idx_range(const YAML::Node &node, std::map<
           {
             single_calib_ = calib;
-            break;
+            break; // BUG?: The break applies to the switch, but should apply to the for loop here?
```
Consider `!single_calib_` in loop as well.

rhaschke commented:
Dear @guihomework, I addressed the discussed issues and force-pushed to have a clean history. 
When doing the final review you might want to compare to the previous state (before the force-push).
If your testing passes as well, this should be ready to merge. I left out the verbosity commit.

rhaschke commented:
The log level cannot directly be set in a launch file, unfortunately. The only option is to provide a config file (in the package), which is configured via env: http://wiki.ros.org/rosconsole#Configuration
Alternatively, we can set the log level easily in C++ code, based on the ROS parameter `verbose`:
http://wiki.ros.org/rosconsole#Changing_Logger_Levels

Which method do you prefer?

guihomework commented on tactile_merger/CMakeLists.txt:
```diff
@@ -10,30 +10,30 @@ find_package(Eigen3 REQUIRED)
 
-add_executable(${PROJECT_NAME} src/merger_node.cpp)
-target_link_libraries(${PROJECT_NAME} lib${PROJECT_NAME})
+add_executable(merger_node src/merger_node.cpp)
```
I would suggest to do the rename in a separate PR since it will affect our launch files in other packages, whereas the rest of the PR won't.
and since we will have to edit our wiki doc + launch files, I would also like to take the opportunity to rename this project or at least the node to something that is not called **merger**.  

The reason is that during recent explanation/tutorial to a colleague, it was very confusing to tell that the _tactile_state_publisher_ does "merging" of the various sources of _tactile_states_ into a single topic, that is than fed into a tactile_merger... that does not do merging exactly, more a conversion of group of taxel data to another type. I would prefer some wording including "convert" (since in converts from tactile_state to tactile_contact_state) or aggregate.

rhaschke replied:
New name: `contact_force_estimator`?

guihomework commented on tactile_calibration/tactile_calibration_tools/src/tactile_calibration_tools/calibration_utils.py:
```diff
@@ -5,7 +5,6 @@
 import matplotlib.pyplot as plt
 from datetime import datetime
 import os
-#pip install pwlf
```
I add this comment again as part of the review. Could we not keep that ? it was not left there by mistake. It helps co-workers...

rhaschke replied:
This type of doc doesn't belong into code in my opinion. You have (correctly) mentioned that in the [README.md](https://github.com/ubi-agni/tactile_toolbox/pull/2/files#diff-c4eb798bc949c4afdb28f8da46a92d82R43).

guihomework replied:
Well, we need to change CMakeLists.txt to install the readme.  Usage is part of the help of the 2 apps, so readme was not installed (and is rarely in our packages), but the functions will not even show the help if this lib is missing. What is the common procedure on Readme for packages in ROS ?

rhaschke replied:
The readme is shown as soon you open the github page. This should be the starting point for all users, shouldn't it?
I don't see a need to install the readme. Where would you want to install it?

guihomework replied:
No idea where to install a readme, this is why I ask. The customers will follow specialized instructions provided via a product manual so should have the info already. It seemed convenient to have it in the code that is the only thing a user easily has access to when installed. I would not do this for a system available package (for which a rosdep exists), here was a particular problem for which the dependency is not clearly appearing in other places (like package.xml).
Can we make the import optional, and warn the user to install this if missing ?

guihomework commented:
> The log level cannot directly be set in a launch file, unfortunately. The only option is to provide a config file (in the package), which is configured via env: http://wiki.ros.org/rosconsole#Configuration
> Alternatively, we can set the log level easily in C++ code, based on the ROS parameter `verbose`:
> http://wiki.ros.org/rosconsole#Changing_Logger_Levels
> 
> Which method do you prefer?

Unfortunately you confirm what I was 90% sure of (as I checked myself not too long ago but might have missed something)
I think only the first method (ENV) permits to set the log level for a named topic which would be a way to not lose the 2-level verbosity I introduced.
If both permit to do it, I still prefer the first one, because one could have 3 config files, for each level of verbosity, then do a group if statement reading a verbose arg in the launch file, and load the appropriate config file. Seems like lot of overhead, but well, if we need to use ROS logging and not print outs, then we are stuck to this overhead.

rhaschke commented:
> I think only the first method (ENV) permits to set the log level for a named topic...

Both methods will work in principle. The first method kind of "teaches" users how to change log levels in general.
The second method hides all these details from the user. So, my main question was: Do you want to teach the user or to simplify things for him?

guihomework commented:
> > I think only the first method (ENV) permits to set the log level for a named topic...
> 
> Both methods will work in principle. The first method kind of "teaches" users how to change log levels in general.
> The second method hides all these details from the user. So, my main question was: Do you want to teach the user or to simplify things for him?

Either I did not understand the first method, or you did not intend to do a "if" in the launch that handles a verbose arg. For me both methods hide the exact loglevel settings. I don't want the user to need to edit and modify a file to get the debug, it should be possible from command line. A minimum of one level should be available this way (debug true or false) if verbose is too complex.

rhaschke commented on tactile_calibration/tactile_calibration_tools/src/tactile_calibration_tools/calibration_utils.py:
```diff
@@ -5,7 +5,6 @@
 from datetime import datetime
 import os
-#pip install pwlf
 import pwlf
```
```suggestion
try:
  import pwlf
except ImportError:
  print "Failed to import pwlf. Ensure that it is installed according to the instructions here: https://github.com/cjekel/piecewise_linear_fit_py#installation"
```

rhaschke commented:
Log level names: ROS_PACKAGE_NAME.detail

rhaschke commented:
@guihomework, I think I addressed the issues discussed offline. If you are fine now (and have approved), I will do some final history cleanup and merge.
